include/veriblock/storage/block_repository.hpp

```diff
@@ -0,0 +1,164 @@
+#ifndef ALT_INTEGRATION_INCLUDE_VERIBLOCK_STORAGE_BLOCK_REPOSITORY_HPP_
+#define ALT_INTEGRATION_INCLUDE_VERIBLOCK_STORAGE_BLOCK_REPOSITORY_HPP_
+
+#include <cstdint>
+#include <memory>
+#include <vector>
+#include <veriblock/slice.hpp>
+
+namespace VeriBlock {
+
+namespace db {
+/**
+ * @class UnrecoverableError
+ * @brief Application is not able to recover from the error occurred in the
+ * storage.
+ *
+ * Example: database is corrupted and can not be opened.
+ *
+ * It is up to implementation to decide a full list of such error types.
+ */
+struct UnrecoverableError : public std::exception {};
+
+/**
+ * @class Error
+ * @brief Application in normal condition can recover from this type of error.
+ *
+ * Example: out of disk space, IO error.
+ *
+ * It is up to implementation to decide a full list of such error types.
+ */
+struct Error : public std::exception {};
+}  // namespace db
+
+/**
+ * @class BlockRepository
+ *
+ * @brief Represents a block tree stored on disk.
+ *
+ * Multiple blocks may be stored at the given height.
+ *
+ * @tparam Block Stored Block type
+ *
+ * @invariant any instance is in valid state - database is opened when valid
+ * instance of BlockRepository created. To close database, delete the instance.
+ */
+template <typename Block>
+struct BlockRepository {
+  //! stored block type
+  using stored_block_t = Block;
+  //! block has type
+  using hash_t = typename Block::hash_t;
+  //! block height type
+  using height_t = typename Block::height_t;
+
+  virtual ~BlockRepository() = default;
+
+  /**
+   * @class WriteBatch
+   * @brief Efficiently implements bulk write operation for BlockRepository.
+   *
+   * @invariant WriteBatch is always in valid state.
+   * @invariant WriteBatch does not modify on-disk storage after put/remove
+   * operations. It does, when \p BlockRepository::commit is executed on this
+   * batch.
+   */
+  struct WriteBatch {
+    virtual ~WriteBatch() = default;
+
+    /**
+     * Write a single block. If block with such hash exists, db will overwrite
+     * it.
+     * @param block to be written in a batch
+     * @return true if we overwrite existing block, false otherwise.
+     */
+    virtual bool put(const stored_block_t& block) = 0;
+
+    /**
+     * Remove a single block from storage identified by its hash.
+     * @param hash block hash
+     * @return true if removed, false if no such element found.
+     */
+    virtual bool removeByHash(const hash_t& hash) = 0;
+
+    /**
+     * Remove potentially many blocks at given height.
+     * @param height block height
+     * @return true if at least one block removed, false otherwise.
+     */
+    virtual bool removeByHeight(height_t height) = 0;
```

Comment:
probably should return int = the removed block count

Reply:
ok not it, but probably something compatible with vector sizes and unsigned like size_t